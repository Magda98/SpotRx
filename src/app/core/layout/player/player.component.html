@if(player) {
<div class="player">
  <ng-container>
    @if(playerService.playbackState | async; as position){
    <mat-slider class="song-slider" min="1" [max]="player.duration" step="10">
      <input (dragEnd)="setPosition($event)" [value]="position" matSliderThumb>
    </mat-slider>
    }
    <div class="song-info">
      <img [alt]="player.track_window.current_track.album.name" class="album-image"
        [src]="player.track_window.current_track.album.images[0].url">
      <div class="album-artist">
        <h2 class="name">{{player.track_window.current_track.name}}</h2>
        <span class="album-name">
          {{player.track_window.current_track.album.name}}
        </span>
        <div class="artist-container">
          @for(item of player.track_window.current_track.artists; track item.name){
          <span class="artist">{{item.name}}</span>
          }
        </div>
      </div>
    </div>
    <div class="action-buttons-container">
      <div class="action-buttons">
        <button appButton type="button" aria-label="previous" [mode]="'text'" (click)="playerService.skipPrev()">
          <mat-icon>skip_previous</mat-icon>
        </button>
        <button appButton type="button" aria-label="play" [mode]="'text'" (click)="playerService.tooglePlay()">
          @if(player.paused){
          <mat-icon>play_circle_outline</mat-icon>
          }
          @else{
          <mat-icon>pause_circle_outline</mat-icon>
          }
        </button>
        <button appButton type="button" aria-label="next" [mode]="'text'" (click)="playerService.skipNext();">
          <mat-icon>skip_next</mat-icon>
        </button>
      </div>
    </div>
    <div class="slider-container">
      <button appButton type="button" [mode]="'text'">
        <app-icon [ngClass]="{'primary': player.shuffle }" (click)="toggleShuffle()" class="icon"
          type="shuffle"></app-icon>
      </button>
      <button appButton type="button" [mode]="'text'" (click)="toggleSavedTrack()">
        <app-icon [ngClass]="{'primary': isSavedTrack() === true }" class="icon" type="heart"></app-icon>
      </button>
      <mat-slider class="volume-slider" min="0" max="1" step="0.1">
        <input (dragEnd)="setVolume($event)" value="0.5" matSliderThumb>
      </mat-slider>
    </div>
  </ng-container>
</div>
}