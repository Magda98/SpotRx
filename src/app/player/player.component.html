<div class="player">
    <!-- <mat-slider min="1" max="5" step="0.5" value="1.5"></mat-slider> -->
    <button *ngIf="playerState | async as player" class="play" (click)="togglePlay()">
        <mat-icon *ngIf="player.paused">play_circle_outline</mat-icon>
        <mat-icon *ngIf="!player.paused">pause_circle_outline</mat-icon>
        {{player.position}}{{player.paused}}
    </button>

</div>