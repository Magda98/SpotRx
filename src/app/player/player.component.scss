.player {
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 140px;
  background-color: #2c2f31;
  padding-left: 0;
  box-sizing: border-box;
  display: grid;
  grid-template-columns: 1fr;
  padding-top: 16px;

  @media screen and (min-width: 980px) {
    grid-template-columns: 1fr 1fr 1fr;
    padding-left: 220px;
  }

  .song-slider {
    position: absolute;
    top: -23px;
    left: 0;
    right: 0;

    @media screen and (min-width: 980px) {
      left: 212px;
    }
  }

  .action-buttons {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-left: 0;
    padding: 0 8px;

    @media screen and (min-width: 980px) {
      margin-left: 25%;
      width: 100%;
      max-width: 250px;
    }
  }

  .song-info {
    display: flex;
    align-items: center;
    padding: 0 8px;

    .album-image {
      width: 60px;
      height: 60px;
      object-fit: cover;
      border-radius: 5px;
      display: block;
      margin-right: 10px;

      @media screen and (min-width: 980px) {
        width: 80px;
        height: 80px;
      }
    }

    .name {
      margin-bottom: 0;
    }

    .artist-container {
      display: flex;
      opacity: 0.6;
      margin-top: 5px;
    }

    .artist {
      margin-right: 5px;
    }

    .album-artist {
      display: flex;
      flex-direction: column;
    }

    .album-name {
      opacity: 0.6;
      font-size: 12px;

      @media screen and (min-width: 980px) {
        font-size: 16px;
      }
    }
  }

  .slider-container {
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .volume-slider {
    display: none;

    @media screen and (min-width: 980px) {
      display: block;
      max-width: 150px;
      width: 100%;
    }
  }
}
