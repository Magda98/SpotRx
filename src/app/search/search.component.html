<div class="search">

    <mat-form-field class="search-input-field" appearance="fill">
        <input type="text" matInput [formControl]="searchFromControl" placeholder="Search...">

    </mat-form-field>
</div>