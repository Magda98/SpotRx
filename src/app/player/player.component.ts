import { PlayerService } from './../services/player.service';
import { Component, OnDestroy, OnInit } from '@angular/core';
import { MatSliderDragEvent } from '@angular/material/slider';
import {
  EMPTY,
  Subscription,
  distinctUntilChanged,
  finalize,
  map,
  switchMap,
  tap,
} from 'rxjs';
import { TrackService } from '../services/track.service';

@Component({
  selector: 'app-player',
  templateUrl: './player.component.html',
  styleUrls: ['./player.component.scss'],
})
export class PlayerComponent implements OnInit, OnDestroy {
  public playerState = this.playerService.getPlayerState();
  public player?: Spotify.PlaybackState;
  public isSavedTrack?: boolean;
  private subscription = new Subscription();
  private toggleTrackSubscription?: Subscription;
  private isFetching = false;

  constructor(
    public playerService: PlayerService,
    private trackService: TrackService
  ) {}

  ngOnInit(): void {
    this.subscription.add(
      this.playerState
        .pipe(
          tap((player) => {
            this.player = player;
          }),
          map((player) => player.track_window.current_track.id),
          distinctUntilChanged(),
          switchMap((id) => {
            if (id) return this.trackService.checkUserSavedTrack(id);
            return EMPTY;
          })
        )
        .subscribe((data) => {
          const [isSavedTrack] = data;
          this.isSavedTrack = isSavedTrack;
        })
    );
  }

  toggleSavedTrack() {
    if (!this.player?.track_window?.current_track?.id) return;

    if (this.isSavedTrack) {
      this.toggleTrackSubscription?.unsubscribe();
      return (this.toggleTrackSubscription = this.trackService
        .deleteTrack(this.player.track_window.current_track.id)
        .subscribe(() => {
          this.isSavedTrack = false;
        }));
    }

    this.toggleTrackSubscription?.unsubscribe();
    return (this.toggleTrackSubscription = this.trackService
      .saveTrack(this.player.track_window.current_track.id)
      .subscribe(() => {
        this.isSavedTrack = true;
      }));
  }

  setVolume(volumeSlider: MatSliderDragEvent) {
    this.playerService.setVolume(volumeSlider.value);
  }

  setPosition(positionSlider: MatSliderDragEvent) {
    this.playerService.setPosition(positionSlider.value);
  }

  toggleShuffle() {
    if (!this.player) return;
    if (this.isFetching) return;

    this.isFetching = true;
    this.player.shuffle = !this.player.shuffle;
    this.subscription.add(
      this.trackService
        .toggleShuffle(this.player.shuffle)
        .pipe(
          finalize(() => {
            this.isFetching = false;
          })
        )
        .subscribe()
    );
  }

  ngOnDestroy(): void {
    this.subscription.unsubscribe();
    this.toggleTrackSubscription?.unsubscribe();
  }
}
