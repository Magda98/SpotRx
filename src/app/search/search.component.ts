import {
  debounceTime,
  distinctUntilChanged,
  filter,
  switchMap,
  EMPTY,
  tap,
} from 'rxjs';
import { Component, OnInit, signal } from '@angular/core';
import { FormControl } from '@angular/forms';
import { TrackService } from '../services/track.service';
import { PlayerService } from '../services/player.service';
import { takeUntilDestroyed, toSignal } from '@angular/core/rxjs-interop';
import { Track } from '../interfaces/track';
import { injectQuery } from '@tanstack/angular-query-experimental';

@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss'],
})
export class SearchComponent {
  public searchFromControl = new FormControl('');
  private search = toSignal(
    this.searchFromControl.valueChanges.pipe(
      debounceTime(300),
      filter((searchValue) => !!searchValue?.length),
      distinctUntilChanged(),
      takeUntilDestroyed()
    )
  );
  searchQuery = injectQuery(() =>
    this.trackService.getSearchResults(this.search() ?? '')
  );
  readonly skeletonLoadingArray = Array.from({ length: 6 }, () => null);

  constructor(
    public trackService: TrackService,
    private playerService: PlayerService
  ) {}

  play(uri: string) {
    this.playerService.playSong([uri], 0);
  }
}
