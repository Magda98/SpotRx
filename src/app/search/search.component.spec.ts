import {
  QueryClient,
  provideAngularQuery,
} from '@tanstack/angular-query-experimental';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from '@testing-library/angular';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { MockBackendInterceptor } from '../../tests/mock-backend.interceptor';
import { SearchComponent } from './search.component';
import { PlayerService } from '../services/player.service';
import userEvent from '@testing-library/user-event';

describe('SearchComponent', () => {
  const renderComponent = () =>
    render(SearchComponent, {
      imports: [HttpClientTestingModule],
      providers: [
        provideAngularQuery(new QueryClient()),
        PlayerService,
        {
          provide: HTTP_INTERCEPTORS,
          useClass: MockBackendInterceptor,
          multi: true,
        },
      ],
    });

  test('should render', async () => {
    const component = await renderComponent();
    const user = userEvent.setup();
    const searchControl = await screen.findByRole('textbox', {
      name: /search/i,
    });
    await user.type(searchControl, 'Maryla');
    await waitForElementToBeRemoved(await screen.findAllByRole('progressbar'));
    component.detectChanges();
    expect(component.container).toMatchSnapshot();
  });
});
