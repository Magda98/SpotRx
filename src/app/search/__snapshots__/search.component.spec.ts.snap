// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent should render 1`] = `
<div
  id="root0"
>
  <div
    class="search"
  >
    <mat-form-field
      appearance="fill"
      class="mat-mdc-form-field search-input-field mat-mdc-form-field-type-mat-input mat-form-field-no-animations mat-form-field-appearance-fill mat-primary ng-untouched ng-valid mat-focused ng-dirty"
    >
      
      <div
        class="mat-mdc-text-field-wrapper mdc-text-field mdc-text-field--filled mdc-text-field--no-label mdc-text-field--focused"
      >
        <div
          class="mat-mdc-form-field-focus-overlay"
        />
        
        <div
          class="mat-mdc-form-field-flex"
        >
          
          
          
          <div
            class="mat-mdc-form-field-infix"
          >
            
            
            
            <input
              aria-invalid="false"
              aria-label="search"
              aria-required="false"
              class="mat-mdc-input-element mat-mdc-form-field-input-control mdc-text-field__input ng-untouched ng-valid cdk-text-field-autofill-monitored ng-dirty"
              matinput=""
              placeholder="Search for specific song..."
              type="text"
            />
          </div>
          
          
        </div>
        <div
          class="mdc-line-ripple mdc-line-ripple--active"
          matformfieldlineripple=""
        />
        
      </div>
      <div
        class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
      >
        <div
          class="mat-mdc-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
          style="opacity: 1; transform: translateY(0%);"
        >
          
          <div
            class="mat-mdc-form-field-hint-spacer"
          />
        </div>
        
        
      </div>
    </mat-form-field>
    <ul
      class="track-list"
    >
      <li
        class="track"
      >
        <div
          class="song-info"
        >
          <img
            class="track-img"
            fetchpriority="auto"
            height="90"
            loading="lazy"
            ng-img="true"
            src="https://i.scdn.co/image/ab67616d00001e02510b37e772bdf2e8204beb0a"
            width="90"
          />
          <div
            class="album-artist"
          >
            <h2
              class="name"
            >
              Małgośka
            </h2>
            <span
              class="album-name"
            >
               Sześć Oceanów (Piosenki 1962-2013) 
            </span>
            <div
              class="time"
            >
              <span
                class="artist"
              >
                Maryla Rodowicz
              </span>
              
            </div>
          </div>
        </div>
        <span
          class="track-artist"
        >
          4:25
        </span>
      </li>
      
    </ul>
    
    
  </div>
</div>
`;
